crates/runtime/src/embeddings/index/retry_client.rs

```diff
@@ -114,7 +117,21 @@ impl S3Vectors for S3VectorRetryClient {
                         | CreateIndexError::ServiceQuotaExceededException(_)
                         | _ => Err(RetryError::permanent(e)),
                     },
-                    SdkError::DispatchFailure(_) => Err(RetryError::transient(e)),
+                    SdkError::DispatchFailure(d) => {
+                        let credentials_not_loaded = d
+                            .as_connector_error()
+                            .and_then(|e| e.source())
+                            .and_then(|s| s.downcast_ref::<CredentialsError>())
+                            .is_some_and(|ce| {
+                                matches!(ce, CredentialsError::CredentialsNotLoaded(_))
+                            });
+
+                        if credentials_not_loaded {
+                            Err(RetryError::permanent(e))
+                        } else {
+                            Err(RetryError::transient(e))
+                        }
+                    }
```

Comment:
Looks like we copy/paste this 16 times. Maybe a macro would be beneficial here.

Reply:
I'm not a huge fan of macros in general - its a much nicer UX imo to be able to debug to the exact line where something happens.

For this specific case its probably fine, but this file is mostly boilerplate code anway.